### PR TITLE
feat(sentrykube): introduce additional possibilities to override config

### PR DESCRIPTION
**Pull request by Litarnus**
This PR adds two possibilities to share common config between cluster configs.

* Hierarchical Sharing: it is possible to introduce an additional directory between `region_overrides` and `region_name` with arbitrary names. This directory can contain a `_values.yaml` file with config values where all values from `region_name/cluster.yaml` are merged into.
* Cluster Sharing: it is possible to add a `_values.yaml` file into `region_overrides/region_name` which works conceptually the same as the hierarchical sharing but it makes more sense if many cluster files exist in a region

### CODE REVIEW

File: libsentrykube/kube.py [hunk 1]
```diff
@@ -111,6 +112,13 @@ def _consolidate_variables(
        overrides are managed manually and they can contain comments. Tools cannot
        preserve comments.
     4. overridden by the cluster file. Which is likely going to be replaced by 2 and 3.
+    5. overridden by creating a hierarchical structure. Adding an intermediate directory
+       in 'region_override' with a '_values.yaml' file allows to have a common config
+       between a set of regions. This is useful if regions in a service are different, but
+       subset of them are similar.
+       It works like point 1 and 2 but with an additional layer in between
+    6. overridden by a common '_values.yaml' within a region folder that has a shared config
+       between all clusters in the region. Values in point 2 will still override those values
```
Comment (fpacifici) [inline libsentrykube/kube.py:121]: I don't think the order is right.

1. Cluster file will be deprecated for service specific override, we are moving to region override only. It still exists because we are in the middle of this process. But if a service has region overrides there should be nothing in the cluster file for that service.
2. managed override (point 3 above) has to be the last override file (assuming the cluster file goes away). 
3. It seems that the common `_values.yaml` file within a region (point 6) should be before point 2 (region override).  If it is shaerd between clusetrs in the same region, the cluster specific file (these - https://github.com/getsentry/ops/tree/master/k8s/services/relay-pop/region_overrides/us) should be more specific than you point 6. 
4. your point 5 should be before 2 as well. 

In summary This should be the order from more generic to more specific
1 The per service value file (this is not region or cluster specific. It is service specific)
5 common override between subset of regions of the same service 
6 common override between clusters in the same region (more specific than 5)
3 cluster specific override 
4 managed override, which is basically equivalent to 3 but managed by tools. 

4 the cluster file is still there for historic reasons, but if you made changes to any of the above the cluster file is supposed to be empty for the specific service.

Comment (Litarnus) [reply libsentrykube/kube.py:121]: Thanks for the detailed comment. You're right that the order isn't correct, I misunderstood the existing comment and thought it's more of a "we have this kind of overrides" instead of them being the order from most to least specific. I will update it accordingly

File: libsentrykube/service.py [hunk 2]
```diff
@@ -112,9 +124,109 @@ def get_service_value_overrides(
...
+        if common_override_values:
+            deep_merge_dict(common_override_values, values)
+            return common_override_values
+
+        return values
     except FileNotFoundError:
-        values = {}
-    return values
+        return {}
+
```
Comment (fpacifici) [inline libsentrykube/service.py:141]: Each of these methods (`get_service_values`, `get_service_value_overrides`, etc.) loads a level of overrides only. Then these are layered in the right order in `_consolidate_variables`. This is intentional so that we separate the logic that loads one level of override from the one that defines the order and make the order apparent.

Please do not merge different levels of overrides in the same method and create a new one instead. The merging should only happen in `_consolidate_variables` only. If you find a lot of duplication feel free to factor that logic out.

File: libsentrykube/service.py [hunk 3]
```diff
@@ -112,9 +124,109 @@ def get_service_value_overrides(
...
+    This function extends the standard region_overrides system by adding support for
+    shared base configurations. This helps reduce duplication across region-specific
+    configurations.
+
+    Directory Structure:
+        region_overrides/
+        └── common_shared_config/      # Arbitrary name for the shared config group
+            ├── _values.yaml           # Base values for this group
+            └── {region_name}/         # Region-specific overrides
+                └── {cluster_name}.yaml # Cluster-specific overrides
```
Comment (fpacifici) [inline libsentrykube/service.py:184]: I am not sure I follow the reasoning here.
1) Which scenario do you want to use the hierarchical value overrides in? Is it about single tenant in order to have some common config for all single tenants vs config common to all saas regions? While I hate the idea that ST should be different I guess that's the world we live into. But if this is not targeting that scenario, could you please elaborate ?
2) If you want to provide a value file that is common to a group of regions why do you have a cluster file here as well? I think in that case the standard cluster specific file (`/region_override/my_cluster.yaml`) should apply rather than having a separate one. In other word I would like this cross regions defaults to apply within the same hierarchy structure we use for all the other services rather than having two parallel hierarchies.

Comment (Litarnus) [reply libsentrykube/service.py:184]: > Which scenario do you want to use the hierarchical value overrides in? Is it about single tenant in order to have some common config for all single tenants vs config common to all saas regions? While I hate the idea that ST should be different I guess that's the world we live into. But if this is not targeting that scenario, could you please elaborate ?

Yes it's for single tenants. I agree that it's not great that the config diverges

> If you want to provide a value file that is common to a group of regions why do you have a cluster file here as well? I think in that case the standard cluster specific file (/region_override/my_cluster.yaml) should apply rather than having a separate one. In other word I would like this cross regions defaults to apply within the same hierarchy structure we use for all the other services rather than having two parallel hierarchies.

Consider the following structure:
```
region_overrides/
├── customer1/
├── customer2/
├── customer3/
├── customer4/
└── customer5/
``` 
`customer1` and `customer2` have similar config and `customer3`, `customer4` and `customer5` have also similar config with enough differences that they can't share a common config with `customer1` and `customer2`. 
As far as I know there is no support for this type of structure which results in a lot of copied config in either the first two or the last 3 customer configs.
Even if we would use `/region_override/my_cluster.yaml`, we can still only share common configs for either of those groups and the other one needs to overwrite it per cluster file.
With my changes we could split it into the following structure where each group can have a single file with a common config:
```
region_overrides/
├── first-group/
│   ├── customer1/
│   └── customer2/
└── second-group/
    ├── customer3/
    ├── customer4/
    └── customer5/
```
Additionally to reducing code duplication we would also see differences between customers more easily.

File: libsentrykube/service.py (hunk 2 above)
Comment (Litarnus) [reply libsentrykube/service.py:141]: Thanks for the review, those are valid points. I will split it up and move the logic

File: libsentrykube/service.py (hunk 3 above)
Comment (fpacifici) [reply libsentrykube/service.py:184]: I'd be fine in having groups. 
Though the override systems must enforce that a given region/cluster must be defined only once. 
Either `us/default.yaml` is in a group or it is in the root `region_override`. It is critical we do not allow both at the same time.


Also please psot this PR in #team-sre-squad-monkeys as this will be an important change that impacts all services and those who manage all services may have opinions.

File: libsentrykube/kube.py [hunk 4]
```diff
@@ -104,26 +107,57 @@ def _consolidate_variables(
...
     # Service defaults from _values
     service_values = get_service_values(service_name, external)
 
     # Service data overrides from services/SERVICE/region_overrides/
     service_value_overrides = get_service_value_overrides(
         service_name, customer_name, cluster_name, external
     )
-    deep_merge_dict(service_values, service_value_overrides)
+
+    if service_value_overrides:
```
Comment (fpacifici) [inline libsentrykube/kube.py:141]: This does not seem right.
This way `services/SERVICE/region_overrides/REGION/_values.yaml` is taken into account only if there is also a cluster specific override file. 

If we go from generic to specific, we should apply `services/SERVICE/region_overrides/REGION/_values.yaml` even if there is no more specific cluster specific region override (`services/SERVICE/region_overrides/REGION/my_cluster.yaml`)

So I think it should be:
```
common_service_values = get_common_regional_override(...)
service_value_overrides = get_service_value_overrides(...)
deep_merge_dict(common_service_values, service_value_overrides)


```

File: libsentrykube/kube.py [hunk 5]
```diff
@@ -104,26 +107,57 @@ def _consolidate_variables(
...
+        if common_service_values is not None:
+            # If a common config file exists then override values with the specific values
+            # and override the less specific ones with the result
+            deep_merge_dict(common_service_values, service_value_overrides)
+            deep_merge_dict(service_values, common_service_values)
+        else:
+            # If no common config exists then just override values with the specific values
+            deep_merge_dict(service_values, service_value_overrides)
+
+    # Service data overrides from services/SERVICE/region_overrides/REGION/_values.yaml
```
Comment (fpacifici) [inline libsentrykube/kube.py:155]: Are you sure this comment is right ?

File: libsentrykube/service.py [hunk 6]
```diff
@@ -1,17 +1,49 @@
...
+    )
+    paths.extend(
+        service_regions_path.glob(
+            f"region_overrides/*/{customer_name}/{cluster_name}.yaml"
+        )
+    )
+
+    if len(paths) > 1:
+        click.echo(f"Expected a single cluster file for customer but got {len(paths)}")
+        click.echo("Found files in:")
```
Comment (fpacifici) [inline libsentrykube/service.py:41]: Please do not print on the console at this level. 
`libsentrykube` is a library and it is independent on how the result is presented.
The cli scripts are outside this package.

Please raise exceptions instead.

File: libsentrykube/kube.py (hunk 4 above)
Comment (Litarnus) [reply libsentrykube/kube.py:141]: I was under the impression that there is always a necessity for a cluster specific override, which I now realise was incorrect.
I will change accordingly

**Review by fpacifici — APPROVED** [review]
